src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,244 @@
+      val firewallRule = new Firewall()
+        .setName(dataprocConfig.clusterFirewallRuleName)
+        .setTargetTags(List(dataprocConfig.clusterNetworkTag).asJava)
+        .setAllowed(List(allowed).asJava)
+
+      val request = new Compute(httpTransport, jsonFactory, getServiceAccountCredential(vmScopes)).firewalls().insert(googleProject, firewallRule)
+
       try {
-        executeGoogleRequest(request)
+        executeGoogleRequest(request) // returns a ComputeOperation
+      } catch {
+        case e: GoogleJsonResponseException => throw new FirewallRuleInsertionException(googleProject, dataprocConfig.clusterFirewallRuleName)
+      }
+    }
+  }
+
+  /* Create init bucket for the cluster and populate it with the cluster initialization files */
+  private def initializeBucket(googleProject: GoogleProject, clusterName: String, bucketName: String): Future[(Bucket, Array[StorageObject])] = {
+    for {
+      // Create a bucket for the cluster
+      bucketResponse <- createBucket(googleProject, clusterName, bucketName)
+      // Add initialization files after the bucket has been created
+      storageObjectsResponse <- initializeBucketObjects(googleProject, clusterName, bucketName)
+    } yield {
+      (bucketResponse, storageObjectsResponse)
+    }
+  }
+
+  /* Create a bucket in the given google project for the initialization files when creating a cluster */
+  private def createBucket(googleProject: GoogleProject, clusterName: String, bucketName:String): Future[Bucket] = {
+    Future {
+      val bucket = new Bucket().setName(bucketName)
+      val bucketInserter = storage.buckets().insert(googleProject, bucket)
+      try {
+        executeGoogleRequest(bucketInserter) // returns a Bucket
       } catch {
-        case e: GoogleJsonResponseException => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Google Request Failed: ${e.getDetails.getMessage}"))
+        case e: GoogleJsonResponseException => throw new BucketNotCreatedException(googleProject, clusterName)
       }
     }
   }
 
+  /* Process the templated cluster init script and put all initialization files in the init bucket */
+  private def initializeBucketObjects(googleProject: GoogleProject, clusterName: String, bucketName: String): Future[Array[StorageObject]] = {
+    Future {
+      // Get the init script
+      val initScriptRaw = scala.io.Source.fromFile(dataprocConfig.configFolderPath + dataprocConfig.initActionsScriptName).mkString
+      // Get a map of the template values to be replaced
+      val replacements = ClusterInitValues(googleProject, clusterName, bucketName, dataprocConfig).toJson.asJsObject.fields
+      // Replace templated values in the raw init script with actual values before populating the bucket
+      val initScript = replacements.foldLeft(initScriptRaw)((a, b) => a.replaceAllLiterally("$(" + b._1 + ")", b._2.toString()))
```

Comment:
can you pull the templating stuff out into a function? i'm hoping it'll end up in workbench-libs fairly soon